src/data/zmod/basic.lean

```diff
@@ -3,526 +3,706 @@ Copyright (c) 2018 Chris Hughes. All rights reserved.
-    exact nat.modeq.modeq_add (nat.mod_mod a n) (nat.mod_mod 1 n) }
+  induction a with a ih, { refl },
+  ext, show (a+1) % (n+1) = fin.val (a+1 : fin (n+1)),
+  { rw [fin.val_add, ← ih, of_nat],
+    exact nat.add_mod _ _ _ }
 end
 
-lemma neg_val' {m : pnat} (n : zmod m) : (-n).val = (m - n.val) % m :=
-have ((-n).val + n.val) % m = (m - n.val + n.val) % m,
-  by { rw [←add_val, add_left_neg, nat.sub_add_cancel (le_of_lt n.is_lt), nat.mod_self], refl },
-(nat.mod_eq_of_lt (fin.is_lt _)).symm.trans (nat.modeq.modeq_add_cancel_right rfl this)
+end fin
+
+/-- The integers modulo `n : ℕ`. -/
+def zmod : ℕ → Type
+| 0     := ℤ
+| (n+1) := fin (n+1)
+
+namespace zmod
+
+instance fintype : Π (n : ℕ) [fact (0 < n)], fintype (zmod n)
+| 0     _ := false.elim $ nat.not_lt_zero 0 ‹0 < 0›
+| (n+1) _ := fin.fintype (n+1)
 
-lemma neg_val {m : pnat} (n : zmod m) : (-n).val = if n = 0 then 0 else m - n.val :=
+lemma card (n : ℕ) [fact (0 < n)] : fintype.card (zmod n) = n :=
 begin
-  rw neg_val',
-  by_cases h : n = 0; simp [h],
-  cases n with n nlt; cases n; dsimp, { contradiction },
-  rw nat.mod_eq_of_lt,
-  apply nat.sub_lt m.2 (nat.succ_pos _),
+  unfreezeI, cases n,
+  { exfalso, exact nat.not_lt_zero 0 ‹0 < 0› },
+  { exact fintype.card_fin (n+1) }
 end
 
-lemma mk_eq_cast {n : ℕ+} {a : ℕ} (h : a < n) : (⟨a, h⟩ : zmod n) = (a : zmod n) :=
-fin.eq_of_veq (by rw [val_cast_nat, nat.mod_eq_of_lt h])
+instance decidable_eq : Π (n : ℕ), decidable_eq (zmod n)
+| 0     := int.decidable_eq
+| (n+1) := fin.decidable_eq _
 
-@[simp] lemma cast_self_eq_zero {n : ℕ+} : ((n : ℕ) : zmod n) = 0 :=
-fin.eq_of_veq (show (n : zmod n).val = 0, by simp [val_cast_nat])
+instance has_repr : Π (n : ℕ), has_repr (zmod n)
+| 0     := int.has_repr
+| (n+1) := fin.has_repr _
 
-lemma val_cast_of_lt {n : ℕ+} {a : ℕ} (h : a < n) : (a : zmod n).val = a :=
-by rw [val_cast_nat, nat.mod_eq_of_lt h]
+instance comm_ring : Π (n : ℕ), comm_ring (zmod n)
+| 0     := int.comm_ring
+| (n+1) := fin.comm_ring n
 
-@[simp] lemma cast_mod_nat (n : ℕ+) (a : ℕ) : ((a % n : ℕ) : zmod n) = a :=
-by conv {to_rhs, rw ← nat.mod_add_div a n}; simp
+instance inhabited (n : ℕ) : inhabited (zmod n) := ⟨0⟩
 
-@[simp, priority 980]
-lemma cast_mod_nat' {n : ℕ} (hn : 0 < n) (a : ℕ) : ((a % n : ℕ) : zmod ⟨n, hn⟩) = a :=
-cast_mod_nat ⟨n, hn⟩ a
+/-- `val a` is a natural number defined as:
+  - for `a : zmod 0` it is the absolute value of `a`
+  - for `a : zmod n` with `0 < n` it is the least natural number in the equivalence class
 
-@[simp] lemma cast_val {n : ℕ+} (a : zmod n) : (a.val : zmod n) = a :=
-by cases a; simp [mk_eq_cast]
+See `zmod.val_min_abs` for a variant that takes values in the integers.
+-/
+def val : Π {n : ℕ}, zmod n → ℕ
+| 0     := int.nat_abs
+| (n+1) := fin.val
 
-@[simp] lemma cast_mod_int (n : ℕ+) (a : ℤ) : ((a % (n : ℕ) : ℤ) : zmod n) = a :=
-by conv {to_rhs, rw ← int.mod_add_div a n}; simp
+lemma val_lt {n : ℕ} [fact (0 < n)] (a : zmod n) : a.val < n :=
```

Comment:
But then type class search will not try to find it automatically later on, right?

Reply:
You mean in this proof, or when this proof is used? If you're talking about when this proof is used, then don't we say that it shouldn't be an instance for the same reason it's not an instance the rest of the time we use `0 < n`?

Reply:
I was talking about when the proof is used. Especially in the case of `zmod`, I think `0 < n` will be a very common assumption that you don't want to think about. In particular for `zmod p` I really like the idea of having this assumption be handled automatically.

Reply:
I think here it should stay like this. I am thinking of the situation where you are working in `zmod p` with a standing assumption `[fact (prime p)]`, and then you would like such nonzero assumptions to be discharged automatically using an instance.

I think it is a good rule of thumb: if you are working with `zmod n` and some assumption on `n` can follow from primality of `n` (or the fact that `n` is prime and odd, which should up often), then it should be a typeclass assumption. If you deal with ring characteristics, also use typeclasses. Otherwise, use regular assumption, unless there is a very good reason for this. Does this seem reasonable to you?

Reply:
Yup, that is basically what I've tried to implement in this refactor.

Reply:
I'm happy to try it out.

Reply:
The let's merge this. You have already done a fantastic work with this, if there are some rough edges we can smoothen them later with more experience.